test-snapshots/snapshots-output/Snapshot.Exports.ss

```diff
@@ -0,0 +1,17 @@
+#!r6rs
+#!chezscheme
+(library
+  (Snapshot.Exports lib)
+  (export
+    TypeAndCtor
+    TypeAndCtor?)
+  (import
+    (prefix (chezscheme) scm:)
+    (prefix (_Chez_Runtime lib) rt:))
+
+  (scm:define TypeAndCtor
+    (scm:quote TypeAndCtor))
+
+  (scm:define TypeAndCtor?
+    (scm:lambda (v)
+      (scm:eq? (scm:quote TypeAndCtor) v))))
```

Comment:
Unexpectedly, `TypeOnly` isn't defined here. Maybe `purescript-backend-optimizer` removes it because it's unused? If that is the case, we could stop that from happening by adding a function to the snapshot that uses a value of that type and exports that function. So, something like
```purs
module Snapshot.Exports
  ( TypeOnly
  , ensureTypeOnlyIsUsed
  , TypeAndCtor(..)
  ) where

data TypeOnly = Ctor String

ensureTypeOnlyIsUsed :: TypeOnly -> String
ensureTypeOnlyIsUsed (Ctor s) = s

data TypeAndCtor = TypeAndCtor
```

Reply:
And that's absolutely broken

```
#!r6rs
#!chezscheme
(library
  (Snapshot.Exports lib)
  (export
    TypeAndCtor
    TypeAndCtor?
    ensureTypeOnlyIsUsed)
  (import
    (prefix (chezscheme) scm:)
    (prefix (_Chez_Runtime lib) rt:))

  (scm:define TypeAndCtor
    (scm:quote TypeAndCtor))

  (scm:define TypeAndCtor?
    (scm:lambda (v)
      (scm:eq? (scm:quote TypeAndCtor) v)))

  (scm:define ensureTypeOnlyIsUsed
    (scm:lambda (v0)
      (Ctor-value0 v0))))
```

The weird thing is that if I try to print the idents of the bindings that are supposed to be translates with

```
bindingsIdents :: Array String                                                                                                                                           
bindingsIdents = map (coerce <<< Tuple.fst) $ Array.concatMap (_.bindings) bindings
```

in `codegenModule`, and somehere later I `trace` that value, I obtain the following

```
[38 of 42] Building Snapshot.Exports
["TypeAndCtor","ensureTypeOnlyIsUsed"]
```

There is no trace (no pun intended) of TypeOnly's Ctor. Maybe it's trying to inline it but we fail to generate the code for it? Or is it a bug in backend-optimizer? I'm confused...

Reply:
I think this is a bug in backend-optimizer... `bindings` come from `usedBindings.value`: https://github.com/aristanetworks/purescript-backend-optimizer/blob/main/src/PureScript/Backend/Optimizer/Convert.purs#L159-L191. Notice how only the identifier is checked? In the above situation, the type and constructor share the same name. Because the constructor isn't exported, that code may also be removing the type as well.